File: spikes/ftps-session-reuse/java/SessionHintSocket.java
import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;
import java.net.InetAddress;
import java.net.Socket;
import java.net.SocketAddress;
import java.net.SocketException;
import java.nio.channels.SocketChannel;

/**
 * Wraps a connected data-channel socket but reports the CONTROL channel's
 * endpoint, so the TLS client session cache is looked up under the control
 * connection's key and offers that session for resumption.
 *
 * Needed because JSSE's BaseSSLSocketImpl.getPort() delegates to the wrapped
 * socket, which would otherwise key the lookup on the data port and miss.
 * Conscrypt (Android) honours the host/port passed to createSocket directly,
 * so this wrapper is harmless there and makes one code path work on both.
 */
final class SessionHintSocket extends Socket {

    private final Socket delegate;
    private final InetAddress hintAddress;
    private final int hintPort;

    SessionHintSocket(Socket delegate, InetAddress hintAddress, int hintPort) {
        this.delegate = delegate;
        this.hintAddress = hintAddress;
        this.hintPort = hintPort;
    }

    // --- the hint: report the control endpoint ---
    @Override public InetAddress getInetAddress() { return hintAddress; }
    @Override public int getPort() { return hintPort; }
    @Override public SocketAddress getRemoteSocketAddress() {
        return new java.net.InetSocketAddress(hintAddress, hintPort);
    }

    // --- everything else goes to the real data socket ---
    @Override public InputStream getInputStream() throws IOException { return delegate.getInputStream(); }
    @Override public OutputStream getOutputStream() throws IOException { return delegate.getOutputStream(); }
    @Override public void close() throws IOException { delegate.close(); }
    @Override public boolean isConnected() { return delegate.isConnected(); }
    @Override public boolean isBound() { return delegate.isBound(); }
    @Override public boolean isClosed() { return delegate.isClosed(); }
    @Override public boolean isInputShutdown() { return delegate.isInputShutdown(); }
    @Override public boolean isOutputShutdown() { return delegate.isOutputShutdown(); }
    @Override public void shutdownInput() throws IOException { delegate.shutdownInput(); }
    @Override public void shutdownOutput() throws IOException { delegate.shutdownOutput(); }
    @Override public InetAddress getLocalAddress() { return delegate.getLocalAddress(); }
    @Override public int getLocalPort() { return delegate.getLocalPort(); }
    @Override public SocketAddress getLocalSocketAddress() { return delegate.getLocalSocketAddress(); }
    @Override public SocketChannel getChannel() { return delegate.getChannel(); }
    @Override public void setSoTimeout(int t) throws SocketException { delegate.setSoTimeout(t); }
    @Override public int getSoTimeout() throws SocketException { return delegate.getSoTimeout(); }
    @Override public void setTcpNoDelay(boolean on) throws SocketException { delegate.setTcpNoDelay(on); }
    @Override public boolean getTcpNoDelay() throws SocketException { return delegate.getTcpNoDelay(); }
    @Override public void setKeepAlive(boolean on) throws SocketException { delegate.setKeepAlive(on); }
    @Override public boolean getKeepAlive() throws SocketException { return delegate.getKeepAlive(); }
    @Override public void setSendBufferSize(int n) throws SocketException { delegate.setSendBufferSize(n); }
    @Override public int getSendBufferSize() throws SocketException { return delegate.getSendBufferSize(); }
    @Override public void setReceiveBufferSize(int n) throws SocketException { delegate.setReceiveBufferSize(n); }
    @Override public int getReceiveBufferSize() throws SocketException { return delegate.getReceiveBufferSize(); }
    @Override public void setSoLinger(boolean on, int l) throws SocketException { delegate.setSoLinger(on, l); }
    @Override public int getSoLinger() throws SocketException { return delegate.getSoLinger(); }
    @Override public void setReuseAddress(boolean on) throws SocketException { delegate.setReuseAddress(on); }
    @Override public boolean getReuseAddress() throws SocketException { return delegate.getReuseAddress(); }
    @Override public void setTrafficClass(int tc) throws SocketException { delegate.setTrafficClass(tc); }
    @Override public int getTrafficClass() throws SocketException { return delegate.getTrafficClass(); }
    @Override public void setOOBInline(boolean on) throws SocketException { delegate.setOOBInline(on); }
    @Override public boolean getOOBInline() throws SocketException { return delegate.getOOBInline(); }
    @Override public void sendUrgentData(int data) throws IOException { delegate.sendUrgentData(data); }
}

File: core-ftp/src/main/kotlin/org/filezilla/ftp/net/SessionHintSocket.kt
package org.filezilla.ftp.net

import java.io.InputStream
import java.io.OutputStream
import java.net.InetAddress
import java.net.InetSocketAddress
import java.net.Socket
import java.net.SocketAddress
import java.nio.channels.SocketChannel

/**
 * Wraps a connected data-channel socket but reports the **control** channel's
 * endpoint, so the TLS client session cache is looked up under the control
 * connection's key and offers that session for resumption.
 *
 * Servers with `require_ssl_reuse` -- vsftpd's default, ProFTPD's `mod_tls`,
 * FileZilla Server -- reject a data connection that does not resume the
 * control session. FileZilla enforces the same rule from its side
 * (`transfersocket.cpp:220-256`).
 *
 * JSSE keys its client session cache on host and port, and takes the port from
 * the *wrapped* socket (`BaseSSLSocketImpl.getPort()` delegates), so passing
 * the control host/port to `createSocket` alone still misses. Conscrypt on
 * Android honours the `createSocket` arguments directly, making this wrapper
 * redundant there but harmless, so one code path works on both.
 *
 * Measured in `spikes/ftps-session-reuse/`: without the wrapper the server
 * reports `SSL_session_reused=False` and answers `522`; with it, `True` and the
 * transfer completes, on both TLS 1.2 and TLS 1.3.
 */
internal class SessionHintSocket(
    private val delegate: Socket,
    private val hintAddress: InetAddress,
    private val hintPort: Int,
) : Socket() {

    // The hint itself.
    override fun getInetAddress(): InetAddress = hintAddress
    override fun getPort(): Int = hintPort
    override fun getRemoteSocketAddress(): SocketAddress = InetSocketAddress(hintAddress, hintPort)

    // Everything else is the real data socket.
    override fun getInputStream(): InputStream = delegate.getInputStream()
    override fun getOutputStream(): OutputStream = delegate.getOutputStream()
    override fun close() = delegate.close()
    override fun isConnected(): Boolean = delegate.isConnected
    override fun isBound(): Boolean = delegate.isBound
    override fun isClosed(): Boolean = delegate.isClosed
    override fun isInputShutdown(): Boolean = delegate.isInputShutdown
    override fun isOutputShutdown(): Boolean = delegate.isOutputShutdown
    override fun shutdownInput() = delegate.shutdownInput()
    override fun shutdownOutput() = delegate.shutdownOutput()
    override fun getLocalAddress(): InetAddress = delegate.localAddress
    override fun getLocalPort(): Int = delegate.localPort
    override fun getLocalSocketAddress(): SocketAddress? = delegate.localSocketAddress
    override fun getChannel(): SocketChannel? = delegate.channel
    override fun setSoTimeout(timeout: Int) { delegate.soTimeout = timeout }
    override fun getSoTimeout(): Int = delegate.soTimeout
    override fun setTcpNoDelay(on: Boolean) { delegate.tcpNoDelay = on }
    override fun getTcpNoDelay(): Boolean = delegate.tcpNoDelay
    override fun setKeepAlive(on: Boolean) { delegate.keepAlive = on }
    override fun getKeepAlive(): Boolean = delegate.keepAlive
    override fun setSendBufferSize(size: Int) { delegate.sendBufferSize = size }
    override fun getSendBufferSize(): Int = delegate.sendBufferSize
    override fun setReceiveBufferSize(size: Int) { delegate.receiveBufferSize = size }
    override fun getReceiveBufferSize(): Int = delegate.receiveBufferSize
    override fun setSoLinger(on: Boolean, linger: Int) = delegate.setSoLinger(on, linger)
    override fun getSoLinger(): Int = delegate.soLinger
    override fun setReuseAddress(on: Boolean) { delegate.reuseAddress = on }
    override fun getReuseAddress(): Boolean = delegate.reuseAddress
    override fun setTrafficClass(tc: Int) { delegate.trafficClass = tc }
    override fun getTrafficClass(): Int = delegate.trafficClass
    override fun setOOBInline(on: Boolean) { delegate.oobInline = on }
    override fun getOOBInline(): Boolean = delegate.oobInline
    override fun sendUrgentData(data: Int) = delegate.sendUrgentData(data)
}
